src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/RoslynCSharpTokenizer.cs

```diff
@@ -386,404 +210,472 @@ private StateResult Data()
+                break;
+            case (int)CSharpSyntaxKind.EqualsToken:
+                TakeTokenContent(token, out content);
+                kind = SyntaxKind.Assign;
+                break;
+            case (int)CSharpSyntaxKind.OpenBracketToken:
+                TakeTokenContent(token, out content);
+                kind = SyntaxKind.LeftBracket;
+                break;
+            case (int)CSharpSyntaxKind.CloseBracketToken:
+                TakeTokenContent(token, out content);
+                kind = SyntaxKind.RightBracket;
+                break;
+            case (int)CSharpSyntaxKind.QuestionToken:
+                TakeTokenContent(token, out content);
+                kind = SyntaxKind.QuestionMark;
+                break;
+            case (int)CSharpSyntaxKind.SemicolonToken:
+                TakeTokenContent(token, out content);
+                kind = SyntaxKind.Semicolon;
+                break;
+            case <= (int)CSharpSyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken and >= (int)CSharpSyntaxKind.TildeToken:
+                TakeTokenContent(token, out content);
+                kind = SyntaxKind.CSharpOperator;
+                break;
+            default:
+                kind = SyntaxKind.Marker;
+                content = Buffer.ToString();
+                Buffer.Clear();
+                break;
+        }
+
+        _currentCSharpTokenTriviaEnumerator = (token.TrailingTrivia.GetEnumerator(), isLeading: false);
+        return Transition(RoslynCSharpTokenizerState.TriviaForCSharpToken, EndToken(content, kind));
+
+        void TakeTokenContent(CSharpSyntaxToken token, out string content)
         {
-            TakeCurrent();
-            return SyntaxKind.GreaterThanEqual;
+            // Use the already-interned string from the C# lexer, rather than realizing the buffer, to ensure that
+            // we don't allocate a new string for every operator token.
+            content = token.ValueText;
+            Debug.Assert(content == Buffer.ToString());
+            Buffer.Clear();
         }
-        return SyntaxKind.GreaterThan;
     }
 
-    private SyntaxKind MinusOperator()
+    private StateResult TokenizedExpectedStringOrCharacterLiteral(CSharpSyntaxKind expectedCSharpTokenKind, SyntaxKind razorTokenKind, string expectedPrefix, string expectedPostfix)
     {
-        if (CurrentCharacter == '>')
-        {
-            TakeCurrent();
-            return SyntaxKind.Arrow;
-        }
-        else if (CurrentCharacter == '-')
+        var curPosition = Source.Position;
+        var result = GetNextResult(NextResultType.Token);
+        var csharpToken = result.Token;
+        // Don't include trailing trivia; we handle that differently than Roslyn
+        var finalPosition = curPosition + csharpToken.Span.Length;
+
+        for (; curPosition < finalPosition; curPosition++)
         {
             TakeCurrent();
-            return SyntaxKind.Decrement;
         }
```

Comment:
This code appears to be duplicated in multiple places in the class, could this be refactored into a private method that can be called from multiple call sites?